File: day5.py
#!/usr/bin/env python3

def split(line, r, target, u, j):
    for l in line:
        if l == target:
            r[1] = ((r[1] - r[0]) // 2) + r[0]
        else:
            r[0] = round((r[1] - r[0]) / 2) + r[0]
    return  r[0] if line[7] == target else r[1]

def seat(line):
    row = split(line, [0, 127], 'F', 0, 7)
    col = split(line, [0, 7], 'L', 7, 9)
    return row * 8 + col

def part1(lines):
    return max([seat(line) for line in lines])

def part2(lines):
    seats = set([seat(line) for line in lines])
    nums = set(range(min(seats), max(seats)+1))
    return list(nums.difference(seats))[0]

if __name__ == "__main__":
    with open("day5.txt") as f:
        lines = [line.strip() for line in f.readlines()]

    assert seat("FBFBBFFRLR") == 357
    ans1 = part1(lines)
    print(ans1)
    assert ans1 == 6714
    ans2 = part2(lines)
    print(ans2)
    assert ans2 == 3435

File: src/day5.rs
use std::fs;
use std::collections::HashSet;

fn split(line: &str, mut r: [i32; 2], target: char, i: usize, j: usize) -> i32 {
    for l in line[i..j].chars() {
        if l == target {
            r[1] = ((r[1] - r[0]) / 2) + r[0];
        }
        else {
            r[0] = ((r[1] as f32 - r[0] as f32) / 2.0).round() as i32 + r[0];
        }
    }
    return if line.chars().nth(j).unwrap() == target { r[0] } else { r[1] };
}

fn seat(line: &str) -> i32 {
    let row = split(line, [0, 127], 'F', 0, 7);
    let col = split(line, [0, 7], 'L', 7, 9);
    return row * 8 + col
}

fn part1(lines: &[&str]) -> i32 {
    return lines.iter().map(|line| seat(line)).max().unwrap();
}

fn part2(lines: &[&str]) -> i32 {
    let seats: Vec<i32> = lines.into_iter().map(|line| seat(line)).collect();
    let seats_set: HashSet<i32> = seats.clone().into_iter().collect();
    let min = seats.clone().into_iter().min().unwrap();
    let max = seats.clone().into_iter().max().unwrap();
    let nums: HashSet<i32> = (min..=max).into_iter().collect();
    return *nums.difference(&seats_set).nth(0).unwrap();
}

pub fn day5() {
    let contents = fs::read_to_string("day5.txt").unwrap();
    let input: Vec<&str> = contents.lines().collect();
    assert_eq!(seat("FBFBBFFRLR"), 357);
    let mut out = part1(&input);
    println!("{}", out);
    assert_eq!(out, 838);
    out = part2(&input);
    println!("{}", out);
    assert_eq!(out, 714);
}